src/wsiToDcm.cpp

```diff
@@ -177,18 +179,63 @@ int WsiToDcm::dicomizeTiff(
     uint32_t row = 1;
     uint32_t column = 1;
     int32_t levelToGet = level;
-    double downsample = 1.;
+    int64_t downsample = 1;
     if (retile) {
       if (downsamples.size() > level && downsamples[level] >= 1) {
-        downsample = downsamples[level];
+        downsample = static_cast<int64_t>(downsamples[level]);
       } else {
-        downsample = ((int64_t)pow(2, level));
+        downsample = static_cast<int64_t>(pow(2, level));
       }
-      levelToGet = openslide_get_best_level_for_downsample(osr, downsample);
+      /*
+        Openslide API  identifies image closest to the downsampled image in size
+        with the API call: openslide_get_best_level_for_downsample(osr,
+        downsample); optimal level selection selects the level with
+        magnification above required level. Downsample acquistions can result in
+        image dimensions which are non-interger multiples of the highest
+        magnification which can result in the openslide_get_level_downsample
+        reporting level downsampeling of a non-fixed multiple:
+
+        Example: Aperio svs imaging,  E.g. (40x -> 10x reports the 10x image has
+        having a downsampeling factor of 4.00018818010427.)
+
+        The code below, computes the desired frame dimensions and then selects
+        the frame which is the best match.
+      */
+      const int64_t tw = firstLevelWidth / downsample;
+      const int64_t th = firstLevelHeight / downsample;
+      for (levelToGet = 1; levelToGet < svs_level_count; ++levelToGet) {
+        int64_t lw, lh;
+        openslide_get_level_dimensions(osr, levelToGet, &lw, &lh);
+        if (lw < tw || lh < th) {
+          break;
+        }
+      }
+      levelToGet -= 1;
     }
+    /*
+        DICOM requires uniform pixel spacing across downsampled image
+        for pixel spacing based metrics to produce images with compatiable
+        coordinate systems across zoom levels.
+
+        Downsampled acquistions can have in image dimensions which are
+       non-interger multiples of the highest magnification. Example: Aperio svs
+       imaging, E.g. (40x -> 10x reports the 10x image has having a
+       downsampeling  factor of 4.00018818010427. This results in non-uniform
+       scaling of the pixels and can result in small, but signifcant
+       mis-alignment in the downsampled imageing. Flooring, the multiplier
+       returned by openslide_get_level_downsample corrects this by restoring
+       consistent downsamping and pixel spacing across the image.
+    */
+    const double multiplicator =
+        floor(openslide_get_level_downsample(osr, levelToGet));
```

Comment:
Is it possible that the integer rounding problem results in a factor that is slightly smaller than the target, e.g. 3.99999 and then calling floor(3.9999) -> 3 when the target was 4?

Reply:
I don't think this is possible.  Downsampling dimensions appear to be = floor (Original  / downsampling factor), or straight integer arithmetic.  This will result in resampled level dimensions which are always <= the exact target multiple (e.g, 9 / 2 = 4 (smaller), 8 / 2 = 4 (exact)).  For these two examples,  openslide_get_level_downsample then would return (9.0/4.0 = 2.25) and (8.0/4.0) = 2.  For openslide_get_level_downsample to be smaller than the target, the level dimensions would need to be calculated as the ceil(Original  / downsampling factor).

Reply:
So if the level downsample is say 2.5 then we round down to 2, this doesn't cause any other accuracy issues elsewhere in the code?